### DISCUSSION TITLE
Any chat that takes longer than 30 seconds to reply fails, very long context chats always fail first time

### DISCUSSION THREAD

**Discussion by matatonic** [open-webui/open-webui#4889] [General]
# Bug Report

## Installation Method

Official docker

## Environment

- **Open WebUI Version:** v0.3.14
- **Ollama (if applicable):** NA

- **Operating System:** Ubuntu 22.04
- **Browser (if applicable):** Firefox 129.0.2 or Chromium 127.0.6533.99 

**Confirmation:**

- [X] I have read and followed all the instructions provided in the README.md.
- [X] I am on the latest version of both Open WebUI and Ollama.
- [ ] I have included the browser console logs.
- [X] I have included the Docker container logs.
- [X] I have provided the exact steps to reproduce the bug in the "Steps to Reproduce" section below.

## Expected Behavior:

A long context chat, or slow operation (such as generating many minutes of speech audio) should continue to work after more than 30 seconds.

## Actual Behavior:

In a long context chat, starting from a cold context cache, a chat request can take a while to warm up the cache, but after 30 seconds without any reply the UI seems to fail and no text is produced. With very large context (100k) even api.groq.com isn't fast enough to start producing text in this time.

## Description

**Bug Summary:**
Any chat (or related action, such as generating speech) which takes more than 30 seconds is aborted.

## Reproduction Details

**Steps to Reproduce:**
Load a 100k context into chat, ask for a reply. 

## Logs and Screenshots

**Browser Console Logs:**
N/A

**Docker Container Logs:**
```
open-webui  | INFO:     192.168.239.14:0 - "GET /api/v1/configs/banners HTTP/1.1" 200 OK
open-webui  | INFO:     192.168.239.14:0 - "GET /api/v1/functions/ HTTP/1.1" 200 OK
open-webui  | INFO:     192.168.239.14:0 - "GET /api/v1/tools/ HTTP/1.1" 200 OK
open-webui  | INFO:     192.168.239.14:0 - "GET /api/v1/prompts/ HTTP/1.1" 200 OK
open-webui  | INFO:     192.168.239.14:0 - "GET /api/v1/chats/tags/all HTTP/1.1" 200 OK
open-webui  | ERROR:    Exception in ASGI application
open-webui  |   + Exception Group Traceback (most recent call last):
open-webui  |   |   File "/usr/local/lib/python3.11/site-packages/starlette/_utils.py", line 87, in collapse_excgroups
open-webui  |   |     yield
open-webui  |   |   File "/usr/local/lib/python3.11/site-packages/starlette/middleware/base.py", line 190, in __call__
open-webui  |   |     async with anyio.create_task_group() as task_group:
open-webui  |   |   File "/usr/local/lib/python3.11/site-packages/anyio/_backends/_asyncio.py", line 680, in __aexit__
open-webui  |   |     raise BaseExceptionGroup(
open-webui  |   | ExceptionGroup: unhandled errors in a TaskGroup (1 sub-exception)
open-webui  |   +-+---------------- 1 ----------------
open-webui  |     | Traceback (most recent call last):
open-webui  |     |   File "/usr/local/lib/python3.11/site-packages/uvicorn/protocols/http/httptools_impl.py", line 401, in run_asgi
open-webui  |     |     result = await app(  # type: ignore[func-returns-value]
open-webui  |     |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
open-webui  |     |   File "/usr/local/lib/python3.11/site-packages/uvicorn/middleware/proxy_headers.py", line 70, in __call__
open-webui  |     |     return await self.app(scope, receive, send)
open-webui  |     |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
open-webui  |     |   File "/usr/local/lib/python3.11/site-packages/fastapi/applications.py", line 1054, in __call__
open-webui  |     |     await super().__call__(scope, receive, send)
open-webui  |     |   File "/usr/local/lib/python3.11/site-packages/starlette/applications.py", line 123, in __call__
open-webui  |     |     await self.middleware_stack(scope, receive, send)
open-webui  |     |   File "/usr/local/lib/python3.11/site-packages/starlette/middleware/errors.py", line 186, in __call__
open-webui  |     |     raise exc
open-webui  |     |   File "/usr/local/lib/python3.11/site-packages/starlette/middleware/errors.py", line 164, in __call__
open-webui  |     |     await self.app(scope, receive, _send)
open-webui  |     |   File "/usr/local/lib/python3.11/site-packages/starlette/middleware/base.py", line 189, in __call__
open-webui  |     |     with collapse_excgroups():
open-webui  |     |   File "/usr/local/lib/python3.11/contextlib.py", line 158, in __exit__
open-webui  |     |     self.gen.throw(typ, value, traceback)
open-webui  |     |   File "/usr/local/lib/python3.11/site-packages/starlette/_utils.py", line 93, in collapse_excgroups
open-webui  |     |     raise exc
open-webui  |     |   File "/usr/local/lib/python3.11/site-packages/starlette/middleware/base.py", line 191, in __call__
open-webui  |     |     response = await self.dispatch_func(request, call_next)
open-webui  |     |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
open-webui  |     |   File "/app/backend/main.py", line 798, in update_embedding_function
open-webui  |     |     response = await call_next(request)
open-webui  |     |                ^^^^^^^^^^^^^^^^^^^^^^^^
open-webui  |     |   File "/usr/local/lib/python3.11/site-packages/starlette/middleware/base.py", line 166, in call_next
open-webui  |     |     raise RuntimeError("No response returned.")
open-webui  |     | RuntimeError: No response returned.
open-webui  |     +------------------------------------
open-webui  | 
open-webui  | During handling of the above exception, another exception occurred:
open-webui  | 
open-webui  | Traceback (most recent call last):
open-webui  |   File "/usr/local/lib/python3.11/site-packages/uvicorn/protocols/http/httptools_impl.py", line 401, in run_asgi
open-webui  |     result = await app(  # type: ignore[func-returns-value]
open-webui  |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
open-webui  |   File "/usr/local/lib/python3.11/site-packages/uvicorn/middleware/proxy_headers.py", line 70, in __call__
open-webui  |     return await self.app(scope, receive, send)
open-webui  |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
open-webui  |   File "/usr/local/lib/python3.11/site-packages/fastapi/applications.py", line 1054, in __call__
open-webui  |     await super().__call__(scope, receive, send)
open-webui  |   File "/usr/local/lib/python3.11/site-packages/starlette/applications.py", line 123, in __call__
open-webui  |     await self.middleware_stack(scope, receive, send)
open-webui  |   File "/usr/local/lib/python3.11/site-packages/starlette/middleware/errors.py", line 186, in __call__
open-webui  |     raise exc
open-webui  |   File "/usr/local/lib/python3.11/site-packages/starlette/middleware/errors.py", line 164, in __call__
open-webui  |     await self.app(scope, receive, _send)
open-webui  |   File "/usr/local/lib/python3.11/site-packages/starlette/middleware/base.py", line 189, in __call__
open-webui  |     with collapse_excgroups():
open-webui  |   File "/usr/local/lib/python3.11/contextlib.py", line 158, in __exit__
open-webui  |     self.gen.throw(typ, value, traceback)
open-webui  |   File "/usr/local/lib/python3.11/site-packages/starlette/_utils.py", line 93, in collapse_excgroups
open-webui  |     raise exc
open-webui  |   File "/usr/local/lib/python3.11/site-packages/starlette/middleware/base.py", line 191, in __call__
open-webui  |     response = await self.dispatch_func(request, call_next)
open-webui  |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
open-webui  |   File "/app/backend/main.py", line 798, in update_embedding_function
open-webui  |     response = await call_next(request)
open-webui  |                ^^^^^^^^^^^^^^^^^^^^^^^^
open-webui  |   File "/usr/local/lib/python3.11/site-packages/starlette/middleware/base.py", line 166, in call_next
open-webui  |     raise RuntimeError("No response returned.")
open-webui  | RuntimeError: No response returned.
open-webui  | ERROR [apps.openai.main] Connection error: 

```

**Screenshots/Screen Recordings (if applicable):**

![image](https://github.com/user-attachments/assets/2c46f334-688c-4cbb-8bf8-e4caead4ba6f)

## Additional Information

My whole setup uses various openai compatible API's, so I don't know if it's related to anything else, or if it's an openai api client issue. The default OpenAI API client timeout should be 10 minutes. I've tried setting:
```
OPENAI_TIMEOUT=600
#and
AIOHTTP_CLIENT_TIMEOUT=600
```

But that doesn't fix anything, still failed after 30 seconds.

**Comment by matatonic** [comment]
Is this not a bug? What does this mean?

**Reply by tjbck** [reply]
To clarify, this error occurs with `/chat/completions` endpoint correct?

**Reply by matatonic** [reply]
Yes.

**Reply by tjbck** [reply]
Hmm, setting `AIOHTTP_CLIENT_TIMEOUT` should've worked. Could you provide a step by step guide on to reproduce the issue?

**Reply by matatonic** [reply]
A very large text context, or a very long conversation always fails for me, like 100,000 tokens. I have a few existing 100,000 token chat's I created, and after restarting tabbyapi and restarting open-webui to ensure the cache is cold, I regenerate the last reply in the conversation, which processes the full 100000 tokens, and usually takes more than 30 seconds to begin streaming a reply. Even if the API server can process 3000 tokens/second, it's still not fast enough to start a reply before 30 seconds expires.

In my testing I've used api.groq.com with llama3-70B-versatile (free api with 128k context), but my typical usage is mistral large v2 or command-r-plus and tabbyapi/exllamav2, which I can make time out with much smaller contexts. Once the cache is 'warm' it works ok, but the first generation (and sometimes second) always fail now.

Another failure case is with openedai-speech (tts) and a very long comment, I find that on my 4090 I can generate the (xtts) audio without issues, but when generating on a slower system (2080) it stops requesting generations after about 30 seconds, which leaves the audio incomplete.

I will try AIOHTTP_CLIENT_TIMEOUT again just to be sure I tested correctly.

**Reply by matatonic** [reply]
I just retested with `AIOHTTP_CLIENT_TIMEOUT=600` in the environment of the docker, and same issue after 30 seconds.
On tabbyapi I see: 
```
tabbyapi  | ERROR:    Chat completion generation cancelled by user.
```

The traceback in open-webui is the same as above in my report.

**Reply by matatonic** [reply]
I realize it's a rather huge test case, is there anything I can do to assist setting it up? llama-3.1-8B or nemo-12B also can have very large context, but large isn't really required, just slow.

**Reply by cheahjs** [reply]
Unable to reproduce with a stub server that either returned HTTP headers and then stalled, or never returning any response at all.

If you run any reverse proxies or load balancers in between Open WebUI and your APIs, you will want to verify your configuration.

<details>
<summary>Stub implementation</summary>

```python
from http.server import BaseHTTPRequestHandler, HTTPServer
import json


class MyHandler(BaseHTTPRequestHandler):
    def do_GET(self):
        print(f"Received GET request for {self.path}")

        if self.path == "/models":
            self.send_response(200)
            self.send_header("Content-type", "application/json")
            self.end_headers()

            response_data = {
                "object": "list",
                "data": [
                    {
                        "id": "stall-test",
                        "object": "model",
                        "owned_by": "test",
                    }
                ],
            }
            self.wfile.write(json.dumps(response_data).encode())
        else:
            # Handle other GET requests (e.g., with a default response)
            self.send_response(200)
            self.send_header("Content-type", "text/plain")
            self.end_headers()
            self.wfile.write(b"Hello, world!")

    def do_POST(self):
        print(f"Received POST request for {self.path}")
        # Send headers immediately
        self.send_response(200)
        self.send_header('Content-type', 'text/event-stream')
        self.send_header('Cache-Control', 'no-cache')
        self.send_header('Connection', 'keep-alive')
        self.end_headers()

        # Wait indefinitely before sending the body
        while True:
            pass

        # This code will never be reached
        self.wfile.write(b" This is a POST request.\n\n")


if __name__ == "__main__":
    server_address = ("", 8000)
    httpd = HTTPServer(server_address, MyHandler)
    print("Starting server on port 8000...")
    httpd.serve_forever()
```
</details>

**Reply by matatonic** [reply]
I don't have anything between them other than... docker proxies? (open-webui and tabbyapi run on the same host).

**Reply by matatonic** [reply]
Did you test against the chat/completions endpoint?

**Reply by matatonic** [reply]
I'll try your stub tomorrow.

**Comment by matatonic** [comment]
> Unable to reproduce with a stub server that either returned HTTP headers and then stalled, or never returning any response at all.
> 
> If you run any reverse proxies or load balancers in between Open WebUI and your APIs, you will want to verify your configuration.
> Stub implementation

You were right! It was a proxy, but not between open-webui and the openai api, but my tls proxy, haproxy has a default timeout of 30 seconds. So it was between the browser and open-web that was getting cut off. I'm sorry I didn't include browser console logs as requested! 

My solution for haproxy, for any in the same situation was:

```
defaults
    log     global
    timeout client 600s
    timeout server 600s
```

Thanks so much for your help debugging!

**Reply by RittikaJindalTR** [reply]
I am facing a similar error, while calling my websocket api, if the response takes more than 30 seconds, it times out, is there a param i cna change to allow this.

**Comment by DevilaN** [comment]
[open-webui.log](https://github.com/user-attachments/files/18382081/open-webui.log)
From time to time I am also encountering similar error.
My setup is based on docker-compose and there is no proxy between ollama and open-webui.
To make things worse, this results in entire webui unusable. I need to restart open-webui container or any further new chats will get stuck no matter if model is llama3.2:1b or anything else which should not cause any lag issue.
